packages/nouns-contracts/contracts/governance/NounsDAOV3Votes.sol

```diff
@@ -0,0 +1,354 @@
+        if (isForVoteInLastMinuteWindow) isDefeatedBefore = ds.isDefeated(proposal);
+
+        if (support == 0) {
+            proposal.againstVotes = proposal.againstVotes + votes;
+        } else if (support == 1) {
+            proposal.forVotes = proposal.forVotes + votes;
+        } else if (support == 2) {
+            proposal.abstainVotes = proposal.abstainVotes + votes;
+        }
+
+        if (
+            // only for votes can trigger an objection period
+            // we're in the last minute window
+            isForVoteInLastMinuteWindow &&
+            // first part of the vote flip check
+            // separated from the second part to optimize gas
+            isDefeatedBefore &&
+            // haven't turn on objection yet
+            proposal.objectionPeriodEndBlock == 0 &&
+            // second part of the vote flip check
+            !ds.isDefeated(proposal)
+        ) {
+            proposal.objectionPeriodEndBlock = proposal.endBlock + ds.objectionPeriodDurationInBlocks;
+            
+            emit ProposalObjectionPeriodSet(proposal.id, proposal.objectionPeriodEndBlock);
+        }
+
+        receipt.hasVoted = true;
+        receipt.support = support;
+        receipt.votes = votes;
+
+        return votes;
+    }
+
+    function castObjectionInternal(
+        NounsDAOStorageV3.StorageV3 storage ds,
+        uint256 proposalId,
+        NounsDAOStorageV3.ProposalState proposalState,
+        address voter
+    ) internal returns (uint96) {
+        require(proposalState == NounsDAOStorageV3.ProposalState.ObjectionPeriod, 'not in objection period');
+        NounsDAOStorageV3.Proposal storage proposal = ds._proposals[proposalId];
+        NounsDAOStorageV3.Receipt storage receipt = proposal.receipts[voter];
+        require(receipt.hasVoted == false, 'NounsDAO::castVoteInternal: voter already voted');
+
+        uint96 votes = receipt.votes = ds.nouns.getPriorVotes(voter, proposalCreationBlock(ds, proposal));
+        receipt.hasVoted = true;
+        receipt.support = 0;
+        proposal.againstVotes = proposal.againstVotes + votes;
+
+        return votes;
+    }
+
+    function _refundGas(uint256 startGas) internal {
+        unchecked {
+            uint256 balance = address(this).balance;
+            if (balance == 0) {
+                return;
+            }
+            uint256 basefee = min(block.basefee, MAX_REFUND_BASE_FEE);
+            uint256 gasPrice = min(tx.gasprice, basefee + MAX_REFUND_PRIORITY_FEE);
+            uint256 gasUsed = min(startGas - gasleft() + REFUND_BASE_GAS, MAX_REFUND_GAS_USED);
+            uint256 refundAmount = min(gasPrice * gasUsed, balance);
+            (bool refundSent, ) = msg.sender.call{ value: refundAmount }('');
+            emit RefundableVote(msg.sender, refundAmount, refundSent);
+        }
+    }
+
+    function proposalCreationBlock(NounsDAOStorageV3.StorageV3 storage ds, NounsDAOStorageV3.Proposal storage proposal)
```

Comment:
@eladmallel maybe rename to `proposalVoteSnapshotBlock` ?

Reply:
fixed